### ISSUE TITLE
Prob with sqlparse.format aligning column ids on alias rather than start of expression

### ISSUE THREAD

**Issue by shimrot** [andialbrecht/sqlparse#207]
Using version 0.1.18 of sqlparse with python3.4 (both windows and linux)

In certain cases, I'm seeing the sqlparse.format results with reintent case the columns to align on the alias rather than the start: 

     ...
     FROM
         (SELECT *,
                 MAX(a) AS 'b',
                           COUNT(d) AS 'ds',
                                       group_concat(a
                                                    ORDER BY a SEPARATOR ',') AS 'as'
          FROM
    ...
 
I'm appending a simple script which shows the issue.  It has a simple query and a subquery which don't show the issue.  The 3rd is more complicated (the original query executes fine, and I hope I didn't mess it up in some way changing the field names, but the formatting issue also occurs in my original sql).  I'm didn't discover what triggers the alignment issue.

### Here is the script:

    #!/usr/bin/env python3
    
    import sqlparse
    
    
    if __name__ == '__main__':
    
        sql = '''
              SELECT *, max(a) as a, count(r) as r, 
                     group_concat(c, order by c separator ',') as cs, 
                     avg(b) as avgb 
              FROM t1 
              GROUP BY x,y,z
              '''
        
        print(sqlparse.format(sql, keyword_case='upper', reindent=True, indent_width=4)) #, right_margin=80 ) 
    
        sql = '''
              SELECT * FROM (
                  SELECT *, max(a) as a, count(r) as r, 
                         group_concat(c, order by c separator ',') as cs, 
                         avg(b) as avgb 
                  FROM t1 
                  GROUP BY x,y,z
              )
              '''
        
        print("---------------------------")
        print(sqlparse.format(sql, keyword_case='upper', reindent=True, indent_width=4)) #, right_margin=80 ) 
    
        sql = '''
            SELECT a, b, c, sum(d) as ds
            FROM (
            
            SELECT * FROM
              (SELECT *, MAX(a) as 'b',
                    COUNT(d) as 'ds',
                    group_concat(a ORDER BY a SEPARATOR ',' ) AS 'as'
                  FROM
                  (SELECT * from t1
                    WHERE status <> 'bad' and
                    (a,b)=('xxx','y') and t like '%'
                    ORDER BY x desc
                  ) sr1
                GROUP BY a, b, c, d, e, f, g, h
              ) t1
              WHERE f = l
            
              ) as t1
            GROUP BY a, b, t
            ORDER BY a, b, t
        ;
        '''
        
        print("---------------------------")
        print(sqlparse.format(sql, keyword_case='upper', reindent=True, indent_width=4)) #, right_margin=80 ) 
    
        

### Here is my output:

$ ./SqlParseFormatASAliginProb.py 

    SELECT *,
           max(a) AS a,
           count(r) AS r,
           group_concat(c,
                        ORDER BY c separator ',') AS cs,
           avg(b) AS avgb
    FROM t1
    GROUP BY x,
             y,
             z
    ---------------------------
    
    SELECT *
    FROM
        ( SELECT *,
                 max(a) AS a,
                 count(r) AS r,
                 group_concat(c,
                              ORDER BY c separator ',') AS cs,
                 avg(b) AS avgb
         FROM t1
         GROUP BY x,
                  y,
                  z )
    ---------------------------
    
    SELECT a,
           b,
           c,
           sum(d) AS ds
    FROM
        ( SELECT *
         FROM
             (SELECT *,
                     MAX(a) AS 'b',
                               COUNT(d) AS 'ds',
                                           group_concat(a
                                                        ORDER BY a SEPARATOR ',') AS 'as'
              FROM
                  (SELECT *
                   FROM t1
                   WHERE status <> 'bad'
                       AND (a,
                            b)=('xxx',
                                'y')
                       AND t LIKE '%'
                   ORDER BY x DESC ) sr1
              GROUP BY a,
                       b,
                       c,
                       d,
                       e,
                       f,
                       g,
                       h ) t1
         WHERE f = l ) AS t1
    GROUP BY a,
             b,
             t
    ORDER BY a,
             b,
             t ;

**Comment by MK-GAN** [comment]
hi shimrot, I encounter this bug too,  have you solve this issue yet?

**Comment by shimrot** [comment]
MK-GAN
Sorry, having been going thru all my email in a while. 
To answer your question, I didn't figure out the reason.  I spent a couple of hrs walking thru code and didn't examine further.  I was hoping someone else might know.

**Comment by vmuriart** [comment]
generated a couple statements that reproduce the issue

```python
bad='''
SeLECT 1 
FROM   (SELECT 1 AS one, 
               2 AS two, 
               3                
        FROM   t) t0
'''


good='''
SELECT 1 
FROM   (SELECT 1 AS one, 
               2 AS two, 
               3
        FROM   t) 
'''

good2='''
SELECT 1 
FROM   (SELECT 1 AS one, 
               2 AS two, 
               3 
        FROM   t) as t0
'''
```

Looks to be caused by [itercycle](https://github.com/andialbrecht/sqlparse/blob/master/sqlparse/engine/grouping.py#L154) not being reset after finding the parenthesis

It can be better seen by running

```python
q=sqlparse.parse(bad)[0]
q._pprint_tree()

q=sqlparse.parse(good2)[0]
q._pprint_tree()
```

which results in:

```python
...bad
 +-7 Identifier '(SELECT 1 A...'
   +-0 Parenthesis '(SELECT 1 A...'
     | 0 Punctuation '('
     | 1 DML 'SELECT'
     | 2 Whitespace ' '
     | 3 Integer '1'
     | 4 Whitespace ' '
     | 5 Keyword 'AS'
     | 6 Whitespace ' '
     +-7 IdentifierList 'one, ...'
       +-0 Identifier 'one'
         | 0 Name 'one'
       | 1 Punctuation ','
       | 2 Whitespace ' ...'
       | 3 Integer '2'
     | 8 Whitespace ' '

...good
+-7 Identifier '(SELECT 1 A...'
   +-0 Parenthesis '(SELECT 1 A...'
     | 0 Punctuation '('
     | 1 DML 'SELECT'
     | 2 Whitespace ' '
     +-3 IdentifierList '1 AS one, ...'
       +-0 Identifier '1 AS one'
         | 0 Integer '1'
         | 1 Whitespace ' '
         | 2 Keyword 'AS'
         | 3 Whitespace ' '
         +-4 Identifier 'one'
           | 0 Name 'one'
       | 1 Punctuation ','
       | 2 Whitespace ' ...'
       +-3 Identifier '2 AS two'
         | 0 Integer '2'
         | 1 Whitespace ' '
         | 2 Keyword 'AS'
         | 3 Whitespace ' '
         +-4 Identifier 'two'
           | 0 Name 'two'
       | 4 Punctuation ','
       | 5 Whitespace ' ...'
       | 6 Integer '3'
     | 4 Whitespace ' '

```

**Comment by vmuriart** [comment]
I fixed this issue in the pr #235 on commit https://github.com/andialbrecht/sqlparse/pull/235/commits/1e770d9b99983fa82375b71c5b3f5d8c6bb1a72e

Below is the new output:

```sql
SELECT *,
       max(a) AS a,
       count(r) AS r,
       group_concat(c,
                    ORDER BY c separator ',') AS cs,
       avg(b) AS avgb
FROM t1
GROUP BY x,
         y,
         z
---------------------------

SELECT *
FROM
    ( SELECT *,
             max(a) AS a,
             count(r) AS r,
             group_concat(c,
                          ORDER BY c separator ',') AS cs,
             avg(b) AS avgb
     FROM t1
     GROUP BY x,
              y,
              z )
---------------------------

SELECT a,
       b,
       c,
       sum(d) AS ds
FROM
    ( SELECT *
     FROM
         (SELECT *,
                 MAX(a) AS 'b',
                 COUNT(d) AS 'ds',
                 group_concat(a
                              ORDER BY a SEPARATOR ',') AS 'as'
          FROM
              (SELECT *
               FROM t1
               WHERE status <> 'bad'
                   AND (a,
                        b)=('xxx',
                            'y')
                   AND t LIKE '%'
               ORDER BY x DESC ) sr1
          GROUP BY a,
                   b,
                   c,
                   d,
                   e,
                   f,
                   g,
                   h ) t1
     WHERE f = l ) AS t1
GROUP BY a,
         b,
         t
ORDER BY a,
         b,
         t ;
```
